fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneGroupByKeysRule.java

```diff
@@ -0,0 +1,208 @@
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.common.Pair;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.AggType;
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalFilterOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalLimitOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+// for aggregate queries with group by keys like `group by col,expr(col),constant` or `group by constant`,
+// these expr and constant won't affect the effect of aggregation grouping, we can remove them
+public class PruneGroupByKeysRule extends TransformationRule {
+    public PruneGroupByKeysRule() {
+        super(RuleType.TF_PRUNE_GROUP_BY_KEYS, Pattern.create(OperatorType.LOGICAL_AGGR)
+                .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT, OperatorType.PATTERN_LEAF)));
+    }
+
+    @Override
+    public boolean check(final OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggOperator = (LogicalAggregationOperator) input.getOp();
+        List<ColumnRefOperator> groupingKeys = aggOperator.getGroupingKeys();
+        if (groupingKeys == null || groupingKeys.isEmpty()) {
+            return false;
+        }
+        return true;
+    }
+
+    private boolean hasFilter(final OptExpression input) {
+        // child of project operator
+        Operator childOperator = input.getInputs().get(0).getInputs().get(0).getOp();
+        if (childOperator instanceof LogicalFilterOperator) {
+            return true;
+        }
+        if (childOperator instanceof LogicalScanOperator) {
+            LogicalScanOperator scanOperator = (LogicalScanOperator) childOperator;
+            if (scanOperator.getPredicate() != null) {
+                return true;
+            }
+        }
```

Comment:
check `scanOperator.getPredicate() != null` ? not only scan/filter, any operator support predicate

Reply:
> check `scanOperator.getPredicate() != null` ? not only scan/filter, any operator support predicate

this is a silly change, it's unreasonable to rely on predicate when rewriting. I have remove it